packages/labs/ssr/src/lib/element-renderer.ts

```diff
@@ -8,16 +8,77 @@
 
 export type Constructor<T> = {new (): T};
 
+import {createRequire} from 'module';
+const require = createRequire(import.meta.url);
+
 // eslint-disable-next-line @typescript-eslint/no-var-requires
 const escapeHtml = require('escape-html') as typeof import('escape-html');
 
 import {RenderInfo} from './render-lit-html.js';
 
+export type ElementRendererConstructor = (new (
+  tagName: string
+) => ElementRenderer) &
+  typeof ElementRenderer;
+
+type AttributesMap = Map<string, string>;
+
+export const getElementRenderer = (
+  {elementRenderers}: RenderInfo,
+  tagName: string,
+  ceClass: typeof HTMLElement = customElements.get(tagName),
+  attributes: AttributesMap = new Map()
+): ElementRenderer | undefined => {
+  if (ceClass === undefined) {
+    console.warn(`Custom element ${tagName} was not registered.`);
+    return;
+  }
+  // TODO(kschaaf): Should we implement a caching scheme, e.g. keyed off of
+  // ceClass's base class to prevent O(n) lookups for every element (probably
+  // not a concern for the small number of element renderers we'd expect)? Doing
+  // so would preclude having cross-cutting renderers to e.g. no-op render all
+  // custom elements with a `client-only` attribute, so punting for now.
+  for (const renderer of elementRenderers) {
+    if (renderer.matchesClass(ceClass, tagName, attributes)) {
+      return new renderer(tagName);
+    }
+  }
+  return undefined;
+};
+
 /**
  * An object that renders elements of a certain type.
  */
 export abstract class ElementRenderer {
-  constructor(public element: HTMLElement) {}
+  element!: HTMLElement;
+  tagName: string;
+
+  /**
+   * Should be implemented to return true when the given custom element class
```

Comment:
We're going to need some kind of precedence mechanism to work with subclassing. We want to match the most specific renderer we can.

Reply:
Discussed and agreed to handle in the future.